src/TDB2.cpp

```diff
@@ -1041,124 +235,128 @@ void TDB2::show_diff (
   if (Context::getContext ().config.getBoolean ("gc"))
   {
-    // Load pending, check whether completed changes size
-    auto size_before = completed._tasks.size ();
-    pending.load_tasks (/*from_gc =*/ true);
-    if (size_before != completed._tasks.size ())
-    {
-      // GC moved tasks from pending to completed
-      pending._dirty = true;
-      completed._dirty = true;
-    }
-    else if (pending._dirty)
-    {
-      // A waiting task in pending was woken up
-      pending._dirty = true;
-    }
-
-    // Load completed, check whether pending changes size
-    size_before = pending._tasks.size ();
-    completed.load_tasks (/*from_gc =*/ true);
-    if (size_before != pending._tasks.size ())
-    {
-      // GC moved tasks from completed to pending
-      pending._dirty = true;
-      completed._dirty = true;
-    }
-
-    // Update blocked/blocking status after GC is finished
-    if (pending._auto_dep_scan)
-      pending.dependency_scan ();
-    if (completed._auto_dep_scan)
-      completed.dependency_scan ();
+    replica.rebuild_working_set (true);
   }
 
   Context::getContext ().time_gc_us += timer.total_us ();
 }
 
-////////////////////////////////////////////////////////////////////////////////
-// Next ID is that of the last pending task plus one.
-int TDB2::next_id ()
-{
-  return _id++;
-}
-
 ////////////////////////////////////////////////////////////////////////////////
 // Latest ID is that of the last pending task.
 int TDB2::latest_id ()
 {
-  return _id - 1;
+  const tc::WorkingSet &ws = working_set ();
+  return (int)ws.largest_index ();
 }
 
 ////////////////////////////////////////////////////////////////////////////////
 const std::vector <Task> TDB2::all_tasks ()
 {
-  std::vector <Task> all (pending._tasks.size () +
-                          pending._added_tasks.size () +
-                          completed._tasks.size () +
-                          completed._added_tasks.size ());
-  all = pending.get_tasks ();
-
-  std::vector <Task> extra (completed._tasks.size () +
-                            completed._added_tasks.size ());
-  extra = completed.get_tasks ();
-
-  for (auto& task : extra)
-    all.push_back (task);
+  auto all_tctasks = replica.all_tasks();
+  std::vector <Task> all;
+  for (auto& tctask : all_tctasks)
+    all.push_back (Task (std::move (tctask)));
 
   return all;
 }
 
 ////////////////////////////////////////////////////////////////////////////////
 const std::vector <Task> TDB2::pending_tasks ()
 {
-  return pending.get_tasks ();
+  const tc::WorkingSet &ws = working_set ();
+  auto largest_index = ws.largest_index ();
+
+  std::vector <Task> result;
+  for (size_t i = 0; i <= largest_index; i++) {
+    auto maybe_uuid = ws.by_index (i);
+    if (maybe_uuid.has_value ()) {
+      auto maybe_task = replica.get_task (maybe_uuid.value ());
+      if (maybe_task.has_value ()) {
+        result.push_back (Task (std::move (maybe_task.value ())));
+      }
+    }
+  }
```

Comment:
This seems like a relatively expensive operation compared to simply reading a file as we were doing before. Has any effort been made to minimize how often this gets called? Should there be a TODO for that?

Reply:
This worried me a little bit, too -- this function is basically loading every pending task from the DB, using a distinct transaction for each one. In general that's not how you're supposed to do things with a DB!

We can definitely optimize things later as we identify performance issues.